tool/teleport/common/agentless.go

```diff
@@ -0,0 +1,525 @@
+		return trace.Wrap(err)
+	}
+
+	fmt.Println("Restarting the OpenSSH daemon")
+	if err := a.restartSSHD(); err != nil {
+		return trace.Wrap(err)
+	}
+	return nil
+}
+
+func (a *agentless) openSSHRotateStageUpdate(ctx context.Context, clf config.CommandLineFlags) error {
+	priv, pub, tlspub, err := ReadKeys(clf)
+	if err != nil {
+		// if the keys are not found do an initial join
+		if trace.IsNotFound(err) {
+			log.Debug("no keys found,attempting join from scratch")
+			return trace.Wrap(a.openSSHInitialJoin(ctx, clf))
+		}
+		return trace.Wrap(err)
+	}
+
+	certs, err := a.register(clf, pub, tlspub)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	openSSHCA, err := a.getOpenSSHCA(ctx, clf.InsecureMode, clf.FIPS, priv, certs)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	if err := os.RemoveAll(clf.OpenSSHKeysBackupPath); err != nil {
+		return trace.Wrap(err)
+	}
+
+	// move the old keys to the backup directory
+	if err := os.Rename(clf.OpenSSHKeysPath, clf.OpenSSHKeysBackupPath); err != nil {
+		return trace.Wrap(err)
+	}
+
+	if err := tryCreateDefaultAgentlesKeysDir(clf.OpenSSHKeysPath); err != nil {
+		return trace.Wrap(err)
+	}
+
+	if err := writeKeys(clf.OpenSSHKeysPath, priv, certs, openSSHCA); err != nil {
+		return trace.Wrap(err)
+	}
+
+	if err := a.restartSSHD(); err != nil {
+		return trace.Wrap(err)
+	}
+
+	return nil
+}
+
+func (a *agentless) openSSHRotateStageRollback(clf config.CommandLineFlags) error {
+	if err := os.RemoveAll(clf.OpenSSHKeysPath); err != nil {
+		return trace.Wrap(err)
+	}
+
+	// move the old keys to the backup directory
+	if err := os.Rename(clf.OpenSSHKeysBackupPath, clf.OpenSSHKeysPath); err != nil {
+		return trace.Wrap(err)
+	}
+
+	if err := a.restartSSHD(); err != nil {
+		return trace.Wrap(err)
+	}
+	return nil
+}
+
+func writeKeys(sshdConfigDir string, private []byte, certs *proto.Certs, openSSHCA []byte) error {
+	if err := os.WriteFile(filepath.Join(sshdConfigDir, teleportKey), private, 0600); err != nil {
+		return trace.Wrap(err)
+	}
+
+	if err := os.WriteFile(filepath.Join(sshdConfigDir, teleportCert), certs.SSH, 0600); err != nil {
+		return trace.Wrap(err)
+	}
+
+	if err := os.WriteFile(filepath.Join(sshdConfigDir, teleportOpenSSHCA), openSSHCA, 0600); err != nil {
+		return trace.Wrap(err)
+	}
+
+	return nil
+}
+
+const sshdConfigSectionModificationHeader = "### Section created by 'teleport join openssh'"
+
+func checkSSHDConfigAlreadyUpdated(sshdConfigPath string) error {
+	contents, err := os.ReadFile(sshdConfigPath)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	if strings.Contains(string(contents), sshdConfigSectionModificationHeader) {
+		return trace.AlreadyExists("not updating %s as it has already been modified by teleport", sshdConfigPath)
+	}
+	return nil
+}
+
+const sshdBinary = "sshd"
+
+func updateSSHDConfig(keyDir, sshdConfigPath string) error {
+	// has to write to the beginning of the sshd_config file as
+	// openssh takes the first occurrence of a setting
```

Comment:
Instead of modifying the original config file, can we generate a new file and just include it in the original `sshd_config`? https://man7.org/linux/man-pages/man5/sshd_config.5.html section Include.
You can then add `GENERATED FILE - DO NOT MOFITY` at the beginning hoping that no one will tamper with them. Removing values should be as simple as regenerating the file with only the required values.

Reply:
> Note that the Debian openssh-server package sets several options as standard in /etc/ssh/sshd_config which are not the default in [sshd(8)](https://manpages.debian.org/unstable/openssh-server/sshd.8.en.html):
> - [Include /etc/ssh/sshd_config.d/*.conf](https://manpages.debian.org/unstable/openssh-server/sshd_config.5.en.html#Include)